README.md
# nuflakbrr-usercard.vercel.app

Create React Github User Card App.

## Authors

Contributors names and contact info

Original scripts by <br />
Rahmat Agung Julians | [@rahmatagungj](https://github.com/rahmatagungj)

Convert scripts from TypeScript to JavaScript by <br />
Naufal Akbar Nugroho | [@kbrnugroho](https://instagram.com/kbrnugroho)

## Version History

- 1.0.0
  - [Initial Release](CHANGELOG.md)


CHANGELOG.md
# Change log

## v1.0.0 — Aug. 21, 2021

Initial release.

**Added:**

- Readme
- Change log
- License


src/pages/Home.js
import React from "react";
import { UserContext } from "../contexts/UserContexts";
import { useHistory } from "react-router-dom";

const Home = () => {
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const username = React.useRef(HTMLInputElement);
  const { setUserData } = React.useContext(UserContext);
  const history = useHistory();

  const handleGetUser = async () => {
    setIsLoading(true);
    const toBeFind = username.current?.value.trim();
    const user = await fetch(`https://api.github.com/users/${toBeFind}`, {
      method: "GET",
    });
    const dataJson = await user.json();
    if (dataJson.message !== "Not Found") {
      setUserData(dataJson);
      history.push(`/${username.current?.value}`);
    } else {
      setIsError(true);
    }
    setIsLoading(false);
  };

  return (
    <div className="bg-gray-900">
      <div className="max-w-3xl px-4 mx-auto sm:px-6 xl:max-w-5xl xl:px-0 flex flex-col h-screen">
        <div className="m-auto">
          <div className="bg-white rounded-lg p-10 justify-center items-center flex flex-col">
            <h1 className="text-gray-800 font-bold uppercase mb-2">Github User Card</h1>
            {isError && <h3 className="text-red-800 font-bold uppercase mb-2 text-sm">Profil tidak ditemukan!</h3>}
            <input type="text" name="username" id="username" placeholder="Input Username" className="bg-gray-300 rounded-lg p-2 border-0 focus:border-0 mb-2" ref={username} />
            <button onClick={() => handleGetUser()} disabled={isLoading} className="bg-blue-700 font-medium shadow rounded-lg p-2 text-gray-100 w-full hover:shadow-2xl hover:bg-blue-800">
              Generate!
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Home;


src/contexts/UserContexts.js
import React from "react";

// interface IUserData {
//   login: string;
//   id: number;
//   node_id: string;
//   avatar_url: string;
//   gravatar_id: string;
//   url: string;
//   html_url: string;
//   followers_url: string;
//   following_url: string;
//   gists_url: string;
//   starred_url: string;
//   subscriptions_url: string;
//   organizations_url: string;
//   repos_url: string;
//   events_url: string;
//   received_events_url: string;
//   type: string;
//   site_admin: boolean;
//   name: string;
//   company?: any;
//   blog: string;
//   location: string;
//   email?: any;
//   hireable: boolean;
//   bio: string;
//   twitter_username: string;
//   public_repos: number;
//   public_gists: number;
//   followers: number;
//   following: number;
//   created_at: string;
//   updated_at: string;
// }

const initialContext = {
  userData: null,
  setUserData: () => {},
};

// interface IUserContext {
//   userData: IUserData | null;
//   setUserData: (userData: null) => void;
// }

export const UserContext = React.createContext(initialContext);

const UserProvider = ({ children }) => {
  const [userData, setUserData] = React.useState(null);
  return <UserContext.Provider value={{ userData, setUserData }}>{children}</UserContext.Provider>;
};

export default UserProvider;


src/pages/UserDetails.js
import React from "react";
import { UserContext } from "../contexts/UserContexts";
import { Redirect } from "react-router-dom";
import { RiGlobalLine, RiTwitterLine, RiUserLocationLine } from "react-icons/ri";

const FlexWithIcons = ({ children }) => {
  return <div className="flex flex-row items-center">{children}</div>;
};

const UserDetails = () => {
  const { userData } = React.useContext(UserContext);

  return (
    <div className="bg-gray-900">
      {!userData && <Redirect to="/" />}
      <div className="max-w-3xl px-4 mx-auto sm:px-6 xl:max-w-5xl xl:px-0 flex flex-col h-screen">
        <div className="m-auto">
          <div className="flex bg-white rounded-lg shadow-xl p-3 m-auto">
            <div className="flex flex-col md:flex-row">
              <img src={userData?.avatar_url} alt={userData?.login} className="w-full rounded-lg" />
              <div className="flex-col p-3 md:p-0 md:px-3 items-center justify-center">
                <a href={userData?.html_url} title={userData?.name} rel="noopener noreferrer">
                  <h1 className="text-gray-800 font-bold capitalize text-lg md:text-xl mb-2 text-center md:text-left">{userData?.name}</h1>
                </a>
                <h2 className="text-gray-800 font-medium capitalize text-md my-2 text-center md:text-left">{userData?.bio}</h2>
                <hr className="my-2" />
                <FlexWithIcons>
                  <RiUserLocationLine />
                  <h3 className="ml-2 text-gray-800 font-medium capitalize text-md my-2 text-center md:text-left">{userData?.location}</h3>
                </FlexWithIcons>
                <a href={userData?.blog} title={userData?.name} rel="noopener noreferrer">
                  <FlexWithIcons>
                    <RiGlobalLine /> <h4 className="ml-2 text-gray-800 font-medium text-md my-2 text-center md:text-left">{userData?.blog.replace("https://", "").toLowerCase()}</h4>
                  </FlexWithIcons>
                </a>
                <FlexWithIcons>
                  <RiTwitterLine />
                  <h5 className="ml-2 text-gray-800 font-medium text-md my-2 text-center md:text-left">{userData?.twitter_username}</h5>
                </FlexWithIcons>

                <div className="flex flex-row items-center my-2 flex-wrap md:flex-nowrap justify-center md:justify-between">
                  <div className="bg-green-700 py-1 px-3 text-gray-100 rounded mr-2 text-center text-sm md:text-md">{userData?.public_repos} Repository</div>
                  <div className="bg-indigo-700 py-1 px-3 text-gray-100 rounded mx-2 text-center text-sm md:text-md">{userData?.followers} Followers</div>
                  <div className="bg-red-600 py-1 px-3 text-gray-100 rounded mx-2 text-center text-sm md:text-md mt-2 md:mt-0">{userData?.following} Following</div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default UserDetails;
